rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,394 @@
+enough how we could implement Benchmark 2 (main menu) using our system.
+Benchmark 1 is irrelevant and Benchmark 3 seems to only require a `Physical`
+layer.
+
+
+## Prior art
+
+I've only had a cursory glance at what already exists in the domain. Beside the
+godot example, I found the [Microsoft
+`FocusManager` documentation](https://docs.microsoft.com/en-us/windows/apps/design/input/focus-navigation-programmatic)
+to be very close to what I am currently doing ([additional Microsoft
+resources](https://docs.microsoft.com/en-us/windows/apps/design/input/focus-navigation))
+
+Our design differs from Microsoft's classical UI navigation in a few key ways.
+We have game-specific assumptions. For example: we assume a "trail" of active elements
+that leads to the sub-sub-submenu we are currently focused in, we rely on that
+trail to navigate containing menus from the focused element in the submenu.
+
+Otherwise, given my little experience in game programming, I am probably overlooking some
+standard practices. My go-to source ("Game Engine Architecture" by Jason
+Gregory) doesn't mention ui navigation systems.
+
+## Rationale and alternatives
+
+### More holistic approach
+
+This system relieves the game developer from implementing themselves an often
+tedious UI navigation system. However, it still asks them to manage the
+graphical aspect of UI such as visual highlights, a more holistic approach
+probably solves this, at the cost of being more complex and less flexible. Such
+a system might be explored in the future, but for the sake of
+having an implementation to work with, I think aiming for a smaller scope is a
+better idea.
+
+### As plugin or integrated in the engine?
+
+I want to first draft this proposition as a standalone plugin ([see on
+github](https://github.com/nicopap/ui-navigation)). The current
+design doesn't require any specific insight into the engine that cannot be
+accessed by a 3rd party plugin. However as I already mentioned, this solves 
+[bevy-specific concerns](https://github.com/bevyengine/bevy/issues/254), and
+bevy would benefit from integrating a good ui navigation system.
+
+### Naming
+
+I'm not comfortable with the names I chose for the various concepts I
+introduced. (such a `Container`, `Plane`, `Physical`) Do not be afraid to
+suggest name changes. Naming is important and I'm already not satisfied with
+the nomenclature I came up with.
+
+## Drawbacks and design limitations
+
+### Unique hierarchy
+
+If we implementation this with a cached navigation tree, it's going to be
+easy to assume accidentally that we have a single fully reachable navigation
+tree. This is not obvious, it's easy to just add the `Container` or `Navigable`
+`Component` to an entity that doesn't have itself a `Container` parent more
+than once.
+
+This needs to be solved, I'm thinking that an error message would be enough in
+this case. But it's worth considering design changes that makes this situation
```

Comment:
IMO this is out of scope for this RFC. This flavor of "graph invariant" is one of the central (and most challenging) problems in relations (a massive ECS-focused effort to better represent entity-entity connections).